Rework filesystem mock

gchazot opened the pull request:
Suggestion of a change in tests description of local filesystem contents to normalise the structure. This allows to remove redundancy of data, and will allow simplification of the code handling filesystem mocking.

Commit:
Rename member variables to explicit internal use

Commit:
Extract method to parse local files and directories in tests

Commit:
Implement new normalised structure to provide mock filesystem data

Fizzadar commented on tests/util.py:
```diff
@@ -229,11 +229,19 @@ def __exit__(self, *args, **kwargs):
+                patch_directories,
+            )
+        else:
+            print(self._parse_files_and_directories(
```
Are these leftovers or to be used when debugging tests (if so can we add a quick comment indicating)?

gchazot replied:
Oops, these are leftovers indeed 59fb3e3

Fizzadar reviewed (commented):
BIG fan of this - a much cleaner way to specify the dirs/files and much less error prone 👍

gchazot committed:
Remove leftover debug logging

gchazot commented:
Cool, when I get the time I'll finalise the migration of all tests and remove the old format

Commit:
Migrate all tests to new local filesystem mock

Commit:
Remove old filesystem mocking format

Commit:
Migrate Windows test to new filesystem mock format

Commit:
Normalise base path for mocked local files.

This is a superficial patch to make sure internal paths to mocked files and directories are using the current OS format.
However, externally, especially in operation parameters, the paths remain in the user-provided format.
This is useful in tests because it means the same JSON can be used on Unix and Windows to describe tests.

gchazot commented:
I've cleaned this up, it's ready now

Fizzadar reviewed (approved):
Thank you for putting this together @gchazot!
